File: sdk/src/main/java/com/meniga/sdk/webservices/requests/GetBudgetParameters.java
package com.meniga.sdk.webservices.requests;

import com.annimon.stream.Optional;
import com.meniga.sdk.helpers.DateTimeUtils;
import com.meniga.sdk.helpers.Joiner;

import org.joda.time.DateTime;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Copyright 2017 Meniga Iceland Inc.
 */
public class GetBudgetParameters extends QueryRequestObject {

    public Long id;
    public List<Long> categoryIds;
    public DateTime startDate;
    public DateTime endDate;
    public Boolean allowOverlappingEntries;
    public Boolean includeEntries;
    public Boolean includeOptionalHistoricalData;

    @Override
    public long getValueHash() {
        int result = id != null ? id.hashCode() : 0;
        result = 31 * result + (categoryIds != null ? categoryIds.hashCode() : 0);
        result = 31 * result + (startDate != null ? startDate.hashCode() : 0);
        result = 31 * result + (endDate != null ? endDate.hashCode() : 0);
        result = 31 * result + (allowOverlappingEntries != null ? allowOverlappingEntries.hashCode() : 0);
        result = 31 * result + (includeEntries != null ? includeEntries.hashCode() : 0);
        result = 31 * result + (includeOptionalHistoricalData != null ? includeOptionalHistoricalData.hashCode() : 0);
        return result;
    }

    @Override
    public Map<String, String> toQueryMap() {
        Map<String, String> queryMap = new HashMap<>();
        Optional.ofNullable(id)
                .map(Object::toString)
                .executeIfPresent(id -> queryMap.put("id", id));

        Optional.ofNullable(categoryIds)
                .executeIfPresent(categoryIds ->
                        queryMap.put("categoryIds", Joiner.join(categoryIds, ",")));

        Optional.ofNullable(startDate)
                .map(DateTimeUtils::toString)
                .executeIfPresent(dateTime -> queryMap.put("startDate", dateTime));

        Optional.ofNullable(endDate)
                .map(DateTimeUtils::toString)
                .executeIfPresent(dateTime -> queryMap.put("endDate", dateTime));

        Optional.ofNullable(allowOverlappingEntries)
                .map(Object::toString)
                .executeIfPresent(value -> queryMap.put("allowOverlappingEntries", value));

        return queryMap;
    }
}

File: sdk/src/main/java/com/meniga/sdk/webservices/requests/GetBudgetParameters.kt
/*
 * Copyright 2017 Meniga Iceland Inc.
 */
package com.meniga.sdk.webservices.requests

import com.meniga.sdk.helpers.DateTimeUtils
import org.joda.time.DateTime

class GetBudgetParameters : QueryRequestObject() {

    var id: Long? = null
    var categoryIds: List<Long>? = null
    var startDate: DateTime? = null
    var endDate: DateTime? = null
    var allowOverlappingEntries: Boolean = false
    var includeEntries: Boolean? = null
    var includeOptionalHistoricalData: Boolean? = null

    override fun getValueHash(): Long {
        var result = if (id != null) id!!.hashCode() else 0
        result = 31 * result + if (categoryIds != null) categoryIds!!.hashCode() else 0
        result = 31 * result + if (startDate != null) startDate!!.hashCode() else 0
        result = 31 * result + if (endDate != null) endDate!!.hashCode() else 0
        result = 31 * result + if (allowOverlappingEntries) 1 else 0
        result = 31 * result + if (includeEntries != null) includeEntries!!.hashCode() else 0
        result = 31 * result + if (includeOptionalHistoricalData != null) includeOptionalHistoricalData!!.hashCode() else 0
        return result.toLong()
    }

    override fun toQueryMap(): Map<String, String> {
        return mutableMapOf<String, String>().also { queryMap ->
            id.toString().let { queryMap["id"] = it }
            categoryIds?.let { queryMap["categoryIds"] = it.joinToString(",") }
            startDate?.let { DateTimeUtils.toString(it) }?.let { queryMap["startDate"] = it }
            endDate?.let { DateTimeUtils.toString(it) }?.let { queryMap["endDate"] = it }
            allowOverlappingEntries.toString().let { queryMap["allowOverlappingEntries"] = it }
        }
    }
}
